### PR TITLE
fix: handle uploaded voice as ref_audio in Voxtral TTS

### PR DESCRIPTION
**Pull request by passionworkeer**
## Summary
When user provides an uploaded speaker voice to Voxtral TTS, resolve it to its reference audio and pass as `ref_audio` parameter instead of as a voice name. Also validates the uploaded voice is audio-backed (not embedding-only) and raises clear errors if reference audio is missing.

## Changes
1. Check if voice is in `uploaded_speakers` dict
2. Validate `embedding_source == "audio"` or `mime_type.startswith("audio/")` — raise `ValueError` if embedding-only
3. Use `self._get_uploaded_audio_data()` to retrieve reference audio — raise `ValueError` if missing
4. Pass `SpeechRequest(input=text, ref_audio=ref_audio)` instead of `voice=...`

Closes #2547

### CODE REVIEW

**Review by lishunyang12 — COMMENTED** [review]
## Review: fix: handle uploaded voice as ref_audio in Voxtral TTS

### Summary

This is a large PR (802 additions, 137 deletions) that, despite the title suggesting a focused Voxtral fix, actually introduces substantial new functionality across multiple TTS backends. The core change (resolving uploaded voices to `ref_audio` for Voxtral) is sound, but the PR bundles in Fish Speech integration, batch speech generation, WAV streaming, speaker embedding upload, and significant refactoring of the shared `create_speech` flow — all in a single commit touching one file.

### What looks good

1. **Voxtral uploaded voice resolution (the stated fix):** The logic in `_build_voxtral_prompt` correctly checks `embedding_source` and `mime_type` to validate audio-backed voices, retrieves the stored audio via `_get_uploaded_audio_data()`, strips the data URI prefix, and passes it as `ref_audio` to `SpeechRequest`. This is well-structured and handles error cases clearly.

2. **Audio duration validation:** Adding `_REF_AUDIO_MIN_DURATION` / `_REF_AUDIO_MAX_DURATION` checks in both `upload_voice` and `_resolve_ref_audio` is a good defensive measure.

3. **`_validate_ref_audio_format` extraction:** De-duplicating the ref_audio format validation into a shared helper method is a clean improvement.

4. **Error messages** are user-friendly and actionable throughout.

### Issues and concerns

**1. Scope creep — PR does far more than the title claims**

The title says "fix: handle uploaded voice as ref_audio in Voxtral TTS" but this PR also adds:
- Full Fish Speech TTS integration (`_build_fish_speech_prompt`, `_validate_fish_tts_request`, `_estimate_fish_prompt_len`, `_estimate_fish_ref_code_len`)
- Batch speech generation (`create_speech_batch`, `_merge_batch_item`, `BatchSpeechRequest`/`BatchSpeechResponse` imports)
- WAV streaming support (`_create_wav_header`, WAV header emission in `_generate_audio_chunks`)
- Speaker embedding upload (`upload_voice_embedding`)
- `_generate_audio_bytes` non-streaming refactor
- `_prepare_speech_generation` orchestration refactor

This makes the PR very hard to review properly. Consider splitting into focused PRs.

**2. `assert` used for runtime validation in `_build_voxtral_prompt`**

```python
assert voice or ref_audio, "Either voice or ref_audio must be provided"
```

`assert` statements are stripped when Python runs with `-O` (optimized mode). This should be a `ValueError` or similar explicit check, especially since this is reachable from user input after validation — a defense-in-depth concern.

**3. Bare `pass` before code in `_validate_qwen_tts_request`**

```python
else:
    # Handle the case where request.voice is NOT None
    pass
    # voice is not None
    voice_lower = request.voice.lower()
```

The `pass` followed by executable code on the next line is confusing. The `pass` is a no-op here and the comment is redundant — this should just have the code directly.

**4. Temp file leak in `_build_fish_speech_prompt`**

```python
with tempfile.NamedTemporaryFile(prefix="fish_ref_", suffix=".npy", delete=False) as f:
    np.save(f, np.asarray(wav_samples, dtype=np.float32))
    ref_audio_path = f.name
```

`delete=False` means these temp files persist indefinitely. There is no cleanup mechanism visible. Under sustained voice-cloning load, this will leak disk space. Consider registering cleanup in a `finally` block or using a managed temp directory.

**5. `_tts_tokenizer` attribute used but never initialized**

In `_build_voxtral_prompt`, `self._tts_tokenizer` is checked and lazily initialized, but I don't see where it's initially set to `None` in `__init__`. If another code path accesses it before `_build_voxtral_prompt` runs, this will raise `AttributeError`. (This may be set elsewhere in the class not visible in the diff — worth confirming.)

**6. Mixed torch/numpy in `_generate_audio_bytes` for cumulative mode**

```python
if async_chunk:
    audio_tensor = torch.cat(non_empty_chunks, dim=-1) if non_empty_chunks else np.zeros(...)
else:
    audio_tensor = np.zeros(...)
    for candidate in reversed(audio_history):
        if candidate.numel() > 0:
            audio_tensor = candidate
            break
```

The fallback for `async_chunk=True` uses `np.zeros` when there are no chunks, but later the code calls `.float().detach().cpu().numpy()` which would fail on an ndarray. The non-async path also calls `.numel()` which is a torch method, not numpy. This suggests the code assumes torch tensors but the fallback creates numpy arrays — a latent type mismatch bug.

**7. `_generate_audio_chunks` — `assert` for sample rate**

```python
assert sr_raw is not None, "First audio chunk must include sample rate metadata for WAV streaming"
```

Same concern as issue #2 — this is runtime validation using `assert`. If the engine ever returns a chunk without `sr`, this would silently pass in optimized mode and produce a WAV header with a possibly wrong default sample rate (24000).

### Nits

- The `import torch` at top level adds startup latency even when serving non-TTS models. Consider lazy import.
- `_batch_max_items` defaults to `getattr(self.engine_client, "tts_batch_max_items", 32)` — this attribute doesn't appear to be documented or set anywhere in the visible code.

### Verdict

The core Voxtral uploaded-voice fix is correct and well-implemented. However, the PR bundles too many unrelated features into a single 800-line diff, making thorough review difficult. I recommend splitting the Fish Speech, batch, WAV streaming, and embedding upload features into separate PRs. The issues noted above (assert usage, temp file leaks, torch/numpy type mismatch) should be addressed.
